provider/osio/osio.go

```diff
@@ -0,0 +1,169 @@
+		safe.Go(func() {
+			select {
+			case <-stop:
+				cancel()
+			}
+		})
+
+		operation := func() error {
+			err := p.fetchToken()
+			if err != nil {
+				return handleCanceled(ctx, err)
+			}
+
+			config, err := p.loadConfig()
+			if err != nil {
+				return handleCanceled(ctx, err)
+			}
+			configChan <- types.ConfigMessage{
+				ProviderName:  providerName,
+				Configuration: config,
+			}
+
+			reload := time.NewTicker(time.Second * time.Duration(p.RefreshSeconds))
+			defer reload.Stop()
+			for {
+				select {
+				case <-reload.C:
+					config, err := p.loadConfig()
+					if err != nil {
+						return handleCanceled(ctx, err)
+					}
+					configChan <- types.ConfigMessage{
+						ProviderName:  providerName,
+						Configuration: config,
+					}
+				case <-ctx.Done():
+					return handleCanceled(ctx, ctx.Err())
+				}
+			}
+		}
+
+		notify := func(err error, time time.Duration) {
+			log.Errorf("%s Provider connection error %+v, retrying in %s", providerName, err, time)
+		}
+		err := backoff.RetryNotify(safe.OperationWithRecover(operation), job.NewBackOff(backoff.NewExponentialBackOff()), notify)
+		if err != nil {
+			log.Errorf("Cannot connect to %s Provider api %+v", providerName, err)
+		}
+	})
+	return nil
+}
+
+func (p *Provider) init(configChan chan<- types.ConfigMessage) {
+	if p.RefreshSeconds <= 0 {
+		p.RefreshSeconds = 60
+	}
+	p.authClient = &authClient{Client: http.DefaultClient}
+}
+
+func (p *Provider) fetchToken() error {
+	if p.tokenResp != nil {
+		return nil
+	}
+	tokenReq := &tokenRequest{GrantType: "client_credentials", ClientID: p.ServiceAccountID, ClientSecret: p.ServiceAccountSecret}
+	tokenResp, err := p.authClient.callTokenAPI(p.TokenAPI, tokenReq)
+	if err != nil {
+		return err
+	}
+	p.tokenResp = tokenResp
+	return nil
+}
+
+func (p *Provider) loadConfig() (*types.Configuration, error) {
+	clusterResponse, err := p.authClient.callClusterAPI(p.ClusterAPI, p.tokenResp)
+	if err != nil {
+		return nil, err
+	}
+	return p.loadRules(clusterResponse), nil
+}
+
+func (p *Provider) loadRules(clusterResp *clusterResponse) *types.Configuration {
+	if len(clusterResp.Clusters) <= 0 {
+		return nil
+	}
+
+	config := &types.Configuration{
+		Frontends: make(map[string]*types.Frontend),
+		Backends:  make(map[string]*types.Backend),
+	}
+
+	defaultBackendExist := false
+	for ind, cluster := range clusterResp.Clusters {
+		if p.defaultBackendURL != "" && p.defaultBackendURL == cluster.APIURL {
+			defaultBackendExist = true
+		}
+		configInd := fmt.Sprintf("%d", ind+1)
+		config.Frontends["frontend"+configInd] = createFrontend(cluster.APIURL, "backend"+configInd)
+		config.Backends["backend"+configInd] = createBackend(cluster.APIURL)
+	}
+	if !defaultBackendExist {
+		p.defaultBackendURL = clusterResp.Clusters[0].APIURL
+		config.Frontends["default"] = createFrontend("default", "default")
+		config.Backends["default"] = createBackend(p.defaultBackendURL)
+	}
```

Comment:
Add default provider to support https://github.com/fabric8-services/fabric8-oso-proxy/blob/master/middlewares/osio/auth.go#L94